Stm32 ccm

lowlander opened the pull request:
Several STM32's have something called CCM Core Coupled Memory. The CCM can only be accessed by the CPU and can not be used for DMA.  But it is fast and the extra 64 kbyte could be useful for things like buffers that aren't meant for IO/DMA.  To use it the patches at 3 new sections ccm_data, ccm_bss and ccm_noinit, that work the same as data (initililized data copied from flash) , bss  (zero initialize data), and noinit (don't initialize data at all). 

The patches work, but touch two things that I don't know that well, linker scripts and device tree files, so I would appreciate it if someone could take a look at it.

lowlander committed:
linker: arm: add stm32 ccm sections

The STM32 has special Core Coupled Memory, ccm for short, that can
only be accessed through the CPU and can not be use for DMA.

The following 3 sections have been added.
- ccm_bss for zero initialized data
- ccm_data for initialized data
- ccm_noinit for uninitialized data

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander committed:
kernel: init: initialize stm32 ccm sections

Initialize the ccm_bss section to zero.
Copy the ccm_data section from the rom section.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander committed:
dts: arm: stm32: Add ccm address and size configurations

Configure ccm size and address via dts. According to the DT
specification, the unit-address of a node must match the
first address specified in the reg property of the node.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -283,6 +286,45 @@ SECTIONS
+        __ccm_bss_start = .;
+        *(.ccm_bss)
+        KEEP(*(SORT_BY_NAME(".ccm_bss.*")))
+        } GROUP_LINK_IN(CCM)
```
I followed the (4 spaces) indenting used in the file. I did notice at some places (not my additions) there was used 8 spaces (for a double indent) and on some other places a TAB was used. 

Maybe add a separate commit to clean up white space of that file ?

lowlander commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -283,6 +286,45 @@ SECTIONS
+        __ccm_data_start = .;
+        *(.ccm_data)
+        KEEP(*(SORT_BY_NAME(".ccm_data.*")))
+        } > CCM AT> ROMABLE_REGION
```
The GROUP_LINK_IN(CCM AT > ROMABEL_REGION)  would work I guess, but that feels like bending the intend of the GROUP_LINK_IN() macro a bit. 

The GROUP_DATA_LINK_IN changes from "> vregion AT> lregion" to "> vregion" depending on XIP or not. Here I always want to "load" it from the ROM to the CCM independent of XIP. 

A GROUP_DATA_LINK_IN like macro would probably be the nicest solution,  suggestions for a name are welcome :-)

erwango reviewed (approved):
Looks good from STM32 point  of view.
One point is that stm32f405.dtsi includes stm32f407.dtsi while it should be the other way round (stm32f407 = stm32f405 + eth + dcmi).
To be corrected in a separate PR, but might be good to fix this first.

galak commented on dts/arm/st/stm32f303.dtsi:
```diff
@@ -5,3 +5,11 @@
 #include <st/stm32f3.dtsi>
+
+/ {
+
```
nit, remove blank line

lowlander replied:
which one, the one before or after the { ?

galak replied:
The one after

galak reviewed (changes requested):
How do we expect to really utilize ccm bss/data?

lowlander commented:
@erwango 

> One point is that stm32f405.dtsi includes stm32f407.dtsi while it should be the other way round (stm32f407 = stm32f405 + eth + dcmi). To be corrected in a separate PR, but might be good to fix this first.

I could add a commit to this PR to fix it, if you are the one that should review it.

lowlander commented:
@galak 

> How do we expect to really utilize ccm bss/data?

`u32_t __ccm_data_section	ccm_data_var = 0x12345678;`
`u32_t __ccm_bss_section	ccm_bss_var;`
`u32_t __ccm_noinit_section	ccm_noinit_var;`

`ccm_data_var` will have the value 0x12345678,  `ccm_bss_var` will be 0 and `ccm_noinit_var` will be what ever was on the memory location.

erwango commented:
> "I could add a commit to this PR to fix it, if you are the one that should review it."
it might be easier indeed, if no one objects

lowlander commented:
@erwango  I looked at it, and there will be some other changes too, like the 407-pinmux (that should be 405-pinmux ?) . So maybe it really is better to do it in a own PR so ppl focus on reviewing just that change.

erwango commented:
@lowlander: Ok sure

erwango commented:
@lowlander: I just pushed #4233. If merged soon, you will just need to rebase on it.

lowlander commented:
@erwango  tanks, I will update this PR tonight.

galak commented:
Can we also work up a test & docs for this feature.  I'm a little concerned about getting use of __ccm_bss_section & __ccm_noinit_section wrong, not sure if there is anything to do about it.

lowlander commented:
@galak I would have liked to have something like;

`int __ccm_section  a = 1234;`

that would put it in the ccm data section, and;

`int __ccm_section b;`

that would put it in the ccm bss section, but I have no idea how to do that automatically.  Because __attribute__((section)) attribute overrules the bss or data the compiler would normally use, so the linker script can't automatically select ccm_data or ccm_bss. 

If someone knows a way to do compile (or link) time checking of wrong use of the attribute((section)) let me know. For example it would be great to get a compile error for `int __ccm_bss b = 123;` because b would be 0 and not 123. 

To the examples/test/docs, yeah I can do that. What would a good place to put documentation ?

lowlander committed:
samples: olimex_stm32_e407: CCM usage example

Add an example program showing the use of the STM32 CCM.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander commented:
@galak  I added an example program
@tbursztyka  I reformatted the linker script

To be honest I am not really happy with the location of the example (boards/olimex_stm32_e407) but it does need a stm32 device. And the drivers dir would also not be the right place because it isn't a driver.  If someone knows a better place, let me know.

dbkinder commented on samples/boards/olimex_stm32_e407/ccm/README.rst:
```diff
@@ -0,0 +1,37 @@
+
+Show usage of the Core Coupled Memory (CCM) that is available
+on several STM32 devices. The very important difference with
+normal RAM is that CCM can not be used for DMA.
```
Maybe add how you can find out how much CCM is available? Are all allocations static?  I presume if you use too much you'l get a linker(?) error. 

Perhaps it would be good to add an introduction to the use of of CCM in the board documentation (e.g., https://www.zephyrproject.org/doc/boards/arm/olimex_stm32_e407/doc/olimex_stm32_e407.html) where you can document using the "__ccm_data_section, __ccm_bss_section, or __ccm_noinit_section" prefixes for using the CCM.

lowlander replied:
"static" as in function-static and file-static (global)  variables, yes.  They just end up in a difference address range (normal RAM is under 0x2000_0000 and CCM under 0x1000_0000).  When you allocate the stack in CCM, you can also use it for stack local variables. But than you can not use stack variables for DMA, which will cause very obscure bugs, that are probably very hard to find. Also AFAIK you can not execute code from CCM, since it is only connected to the databus, but I have found no good docu about that. 

When using up more than there is the linker should complain.  I say "should" because I didn't test it yet (which I will now :-) ) 

Since you are Mr. Documentation, maybe you know a better place for the CCM docu. The CCM is not board specific, it is available on most STM32F4 and STM32F3 cpus. So is there be a place where I can document the CCM in general and than add a reference "This board supports CCM" in the board docu? A general "memory" section (maybe other CPU's have something similar?)  in the docu would be a nice place to really explain the ins and out, like getting a linker error when using to much, that DMA can not work from it, etc.

superna9999 replied:
Yes, It's very generic CCM since is a ST Specific modern way to implement the ARM "Tighly Coupled Memory" that existed on the ARMv6 based Cores like the Cortex-M1.

lowlander replied:
@superna9999  OK I didn't know that, I thought it was a ST32 specific thing.  Any suggestions on how to improve this patch?

dbkinder commented on samples/boards/olimex_stm32_e407/ccm/README.rst:
```diff
@@ -0,0 +1,37 @@
+.. code-block:: console
+
+   $ cd samples/boards/olimex_stm32_e407/ccm
+   $ make
```
Could you also add a "Running" section.  I see the sample writes some output that might be worth showing and explaining here.

lowlander replied:
Good idea, that should be no problem.

dbkinder reviewed (changes requested):
Thanks for adding documentation!  Some suggestions, including documenting CCM in the board documentation.

erwango commented:
@lowlander, @galak, @tbursztyka, @dbkinder, a part from a rebase and cmake rework, what would be missing here?

lowlander commented:
@erwango after @superna9999  comment that this type of memory is not just specific to the STM32 I am wondering if it should be more generic? 

For the rest the test prog output is missing in the docu, I will update that in the weekend.

lowlander commented:
The Cmake switch has changed some things for the DT -> config.h generation it seems.  Currently my patch does not generate CONFIG_CCM_SIZE and CONFIG_CCM_BASE_ADDRESS anymore, and so the compile fails.

I am working on it.

lowlander committed:
scripts/dts/extract_dts_includes.py: add ccm support

Add support for generating CONFIG_CCM_BASE_ADDRESS and CONFIG_CCM_SIZE
from the dts files.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander committed:
dts: add ccm config to olimex stm32 boards

Add ccm config to the Olimex STM32 boards DTS files.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander commented:
Hey all, I had to make some changes after the cmake update.  I now use the DTS files to generate the CONFIG_CCM_* defines.  Please have a good look if things are OK, especially if you already approved it (@erwango ).

erwango commented on dts/arm/yaml/st,stm32-ccm.yaml:
```diff
@@ -0,0 +1,24 @@
+---
```
Would require a rebase as dts/arm/yaml no more exists.
Should be moved under dts/bindings/ccm (/tcm?)

lowlander replied:
I overlooked that the yaml file was now a completely new file, I will look into it tonight after work.

Any comment on the general setup ?  Is it ok like this by doing it with dts and zephyr  choose ccm0 =  &ccm ?

lowlander commented:
@erwango  I moved the yaml file and cleaned up the code a bit. Also enhanced the example.  Let me know what you think.

erwango reviewed (approved):
Fine for me

lowlander commented:
@dbkinder Is the docu ok like this, with the example code output ?

dbkinder commented on samples/boards/olimex_stm32_e407/ccm/README.rst:
```diff
@@ -0,0 +1,186 @@
+   Example end
+
+
+First the CCM sections with their addresse and size are listed, after that
```
change to:

    First, each CCM section is listed with its address and size. Next, some usage examples are shown.
    Note that the ``noinit`` section holds variables with uninitialized data.  After writing to the
    variables, they all should hold the values shown above.

dbkinder commented on samples/boards/olimex_stm32_e407/ccm/README.rst:
```diff
@@ -0,0 +1,186 @@
+holds random data. After writing to the variables they all should hold the
+values shown above.
+
+When the board is reseted without powercycling, the output looks like this;
```
Change to:    

     When the board is reset (without power-cycling), the output looks like this:

dbkinder commented on samples/boards/olimex_stm32_e407/ccm/README.rst:
```diff
@@ -0,0 +1,186 @@
+
+.. code-block:: console
+
+   ***** BOOTING ZEPHYR OS v1.11.0 *****
```
We're not at version 1.11 yet, should this be 1.10.99 for now?

lowlander replied:
@dbkinder  Thank you for taking the time to proofread and the suggestions, I pretty much implemented them 1 to 1. But with the version number , when I now write 1.10.99 it will never be changed again , will it? Wouldn't it than be better to use the version number of the upcoming release (1.11.0 in this case) ?

dbkinder replied:
Well, at least if you run it now, it will show 1.10.99 (if using the master branch) until the 1.11.0 release comes out in March 2018 and then 1.12.0 in June (or so)...

lowlander replied:
So just make it 1.10.99 ?

dbkinder replied:
Yes (my recommendation).

dbkinder commented on samples/boards/olimex_stm32_e407/ccm/README.rst:
```diff
@@ -0,0 +1,186 @@
+
+.. code-block:: console
+
+   ***** BOOTING ZEPHYR OS v1.11.0 *****
```
Same comment here: we're not at version 1.11 yet, should this be 1.10.99 for now?

dbkinder commented on samples/boards/olimex_stm32_e407/ccm/README.rst:
```diff
@@ -0,0 +1,186 @@
+
+
+The difference with the first run is that the ccm_noinit section still has the
+values from the last write. It is important to notice that this is not garanteed,
```
s/garanteed/guaranteed/

dbkinder commented on samples/boards/olimex_stm32_e407/ccm/README.rst:
```diff
@@ -0,0 +1,186 @@
+
+The difference with the first run is that the ccm_noinit section still has the
+values from the last write. It is important to notice that this is not garanteed,
+it still should be considered random left over data.
```
I'd call it "uninitialized leftover data" rather than "random" (since it's not randomly generated data)

dbkinder reviewed (changes requested):
suggested edits

lowlander commented:
@dbkinder I updated the readme following your suggestions, I hope you can give your OK to this PR now.

andrewboie commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -296,6 +299,45 @@ SECTIONS
+	{
+		__ccm_noinit_start = .;
+		*(.ccm_noinit)
+		KEEP(*(SORT_BY_NAME(".ccm_noinit.*")))
```
do we need these KEEP()s here? why are we sorting them? This seems like copypasta.

lowlander replied:
@andrewboie to be honest me ld.scripting knowledge is very limited, so yes it mainly is copy/paste from other sections with comparable functionality.  I just now tried with; 

`*(".ccm_data.*")`

and that seems to work fine too.  Do you want me to remove the KEEP()/SORT() ?

andrewboie replied:
yes.

andrewboie commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -296,6 +299,45 @@ SECTIONS
+		__ccm_start = .;
+		__ccm_bss_start = .;
+		*(.ccm_bss)
+		KEEP(*(SORT_BY_NAME(".ccm_bss.*")))
```
see below

andrewboie commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -296,6 +299,45 @@ SECTIONS
+	{
+		__ccm_data_start = .;
+		*(.ccm_data)
+		KEEP(*(SORT_BY_NAME(".ccm_data.*")))
```
see above

tyrantlizard77 reviewed (commented):
I can't comment on the DTS stuff.
Init changes look OK.
Curious about how the sections are specified in the linker script though. I fear the KEEP() in there may defeat gc-sections

lowlander commented:
@andrewboie  I removed the unneeded KEEP/SORT linker script stuff.

lowlander commented:
@tbursztyka  do you have any change requests for this PR ?  

@dbkinder  is the docu OK like this ?

dbkinder commented on samples/boards/olimex_stm32_e407/ccm/README.rst:
```diff
@@ -0,0 +1,188 @@
+
+For example the olimex STM32 E407 DTS file looks like this:
+
+.. code-block:: console
```
To avoid having the doc go out of sync with the dts file, you can just include it here with (leave off the linenos parameter if you don't want line numbers displayed):

     .. literalinclude:: ../../olimex_stm32_e407.dts
        :linenos:

dbkinder reviewed (changes requested):
Thanks.  One change for the .dts file you pasted in (would be better to use a directive to include it).

lowlander commented:
@dbkinder  docs OK like this ?

dbkinder reviewed (approved):
yep, that works. Thanks.

lowlander commented:
@superna9999 @tbursztyka Any change requests to this PR ?

lowlander commented:
@galak everybody (only don't have feedback from @tbursztyka and @superna9999  ) seem to be OK with this PR, do you have any change requests ?

galak commented on dts/bindings/arm/st,stm32-ccm.yaml:
```diff
@@ -0,0 +1,26 @@
+properties:
+  compatible:
+    type: string
+    category: optional
```
this should be required not optional

galak commented on kernel/init.c:
```diff
@@ -142,6 +142,10 @@ void _bss_zero(void)
 {
 	memset(&__bss_start, 0,
 		 ((u32_t) &__bss_end - (u32_t) &__bss_start));
+#ifdef CONFIG_CCM_BASE_ADDRESS
```
I think this should be moved into _PrepC.

lowlander replied:
@galak  I put it there because the BSS init is also in that location. Both BSS and CCM init have almost the same function, so shouldn't they be close together ?

galak commented on kernel/init.c:
```diff
@@ -162,6 +166,10 @@ void _data_copy(void)
 {
 	memcpy(&__data_ram_start, &__data_rom_start,
 		 ((u32_t) &__data_ram_end - (u32_t) &__data_ram_start));
+#ifdef CONFIG_CCM_BASE_ADDRESS
```
I think this should be moved into _PrepC.

galak reviewed (changes requested):
A few minor things

lowlander committed:
sanitycheck: add ccm_bss, ccm_noinit and ccm_data sections

Add the STM32 ccm_bss, ccm_noinit, abd ccm_data sections
to the list of allowed sections so the sanity check passes.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander commented:
@galak  I fixed the yaml (required vs optional) but I really think the initialization of ccm_bss and ccm_data are in the right place. The initialization of bss, app_bss, data, and app_data are also there. 

Or are you talking about the whole _data_copy() and _bss_zero() functions that should be moved from that file ? 

I think that should be done in a separate PR, because else all discussions start over again and it takes another 4 months to get this PR done.

galak commented:
> @galak I fixed the yaml (required vs optional) but I really think the initialization of ccm_bss and ccm_data are in the right place. The initialization of bss, app_bss, data, and app_data are also there.
> 
> Or are you talking about the whole _data_copy() and _bss_zero() functions that should be moved from that file ?
>

I was suggesting moving them so they are in ARM specific code.

> I think that should be done in a separate PR, because else all discussions start over again and it takes another 4 months to get this PR done.

Sorry, I was hoping to research a better way to accomplish this.
